cljfmt/src/cljfmt/core.cljc

```diff
@@ -675,17 +676,75 @@
+                      (if-some [z (skip-to-linebreak-or-element z)]
+                        (let [at-blank-line? (has-blank-line-after? z)
+                              new-max (if (= col-idx (dec col))
+                                        (max max-p (node-end-position z))
+                                        max-p)
+                              next-col (if (line-break? z) 0 (inc col-idx))]
+                          (if (or at-blank-line? (nil? (z/right* z)))
+                            new-max
+                            (recur (z/right* z) new-max next-col)))
+                        max-p))
+            ;; Apply alignment to current group
+            start-position (inc max-pos)
+            zloc-after (loop [z group-start-zloc, col-idx 0]
+                         (if-some [z (skip-to-linebreak-or-element z)]
+                           (let [at-blank-line? (has-blank-line-after? z)
+                                 z (if (and (= col-idx col) (not (line-break? z)))
+                                     (pad-to-position z start-position)
+                                     z)
+                                 next-col (if (line-break? z) 0 (inc col-idx))]
+                             (if at-blank-line?
+                               z
+                               (if-some [zn (z/right* z)]
+                                 (recur zn next-col)
+                                 z)))
+                           z))]
+        ;; Find next group start (after blank line)
+        (if-let [next-group (loop [z zloc-after]
+                              (when-some [z (skip-to-linebreak-or-element z)]
+                                (if (has-blank-line-after? z)
+                                  (z/right* z)
+                                  (recur (z/right* z)))))]
+          (recur zloc-after next-group)
+          (z/up zloc-after))))
+    zloc))
+
 (defn- align-one-column [zloc col]
   (if-some [zloc (z/down zloc)]
     (let [start-position (inc (max-column-end-position zloc (dec col)))]
       (z/up (edit-column zloc col #(pad-to-position % start-position))))
     zloc))
 
-(defn- align-columns [zloc]
-  (reduce align-one-column zloc (-> zloc z/down count-columns range rest)))
-
-(defn align-map-columns [form]
-  (transform form edit-all z/map? align-columns))
+(defn- column-indexes [zloc]
+  (-> zloc z/down count-columns range rest))
+
+(defn- align-columns
+  ([zloc] (align-columns zloc false))
+  ([zloc blank-lines-separate-alignment?]
+   (if blank-lines-separate-alignment?
+     (reduce (fn [zloc col]
+               (align-one-column-with-groups zloc col))
+             zloc
+             (column-indexes zloc))
+     (reduce (fn [zloc col]
+               (align-one-column zloc col))
+             zloc
+             (column-indexes zloc)))))
+
+(defn align-map-columns [form blank-lines-separate-alignment?]
+  (transform form edit-all z/map? #(align-columns % blank-lines-separate-alignment?)))
```

Comment:
This could be a little more concisely expressed as:

```clojure
(defn- align-columns [zloc alignf]
  (reduce alignf zloc (-> zloc z/down count-columns range rest)))

(defn align-map-columns [form blank-lines-separate-alignment?]
  (let [alignf (if blank-lines-separate-alignment?
                 align-one-column
                 align-one-column-with-groups))]
    (transform form edit-all z/map? #(align-columns % alignf))))
```